### ISSUE TITLE
standardize diagram definitions

### ISSUE THREAD

**Issue by Yokozuna59** [mermaid-js/mermaid#4499]
Most of these rules were discussed in this PR #4486, but it became kind of messy with commits and reviews, so we're going to summarize them here (and update them continuously):

- `.js` files aren't allowed; use `.ts` instead.
- try not to use `// @ts-*` comments, but if you have, please add a description showing the problem. i.g.:

  ```ts
  // @ts-ignore - couldn't cast Document type into HTML
  const html: HTML = docuement;
  ```

## General Diagram Definition

```
packages/mermaid/src/diagrams/
├── parser/
│   └── diagram.jison
├── diagram.spec.ts
├── diagramDb.ts
├── diagramDetector.ts
├── diagramDiagram.ts
├── diagramRenderer.ts
├── diagramStyles.ts
└── diagramTypes.ts
```

- parser/diagram.jison: contains parser and grammar of the diagram.
- diagramDb.ts:
  - fields with there functions, i.g., getters and setters.
  - DEFAULT_DIAGRAM_DB.
  - DEFAUTL_DIAGRAM_CONFIG.
  - title and accessibilities getters and setter.
  - custom clear function (or use common one).
  - config getter.
  - parseDirective.
- diagramDiagram.ts: contains the diagram definition with db, parser, styles, etc.
- diagramDetector.ts:
  - id (diagram keyword).
  - detector.
  - loader.
  - diagram (export the above variables/functions).
- diagramRenderer.ts:
  - draw function.
- diagramStyles.ts: contains css styles for the diagram.
- diagramTypes.ts: contains diagram types and interfaces for DB, styles options, etc.

**Comment by Yokozuna59** [comment]
And shouldn't better to move types in `config.type.ts` into each `diagramTypes.ts`?

**Comment by sidharthv96** [comment]
All the suggestions look great to me.

**Comment by Yokozuna59** [comment]
> All the suggestions look great to me.

Great! Could you add labels to the issue? i.e, `Discussion`.

And do you have something in your mind on behave of separating `setConfig` parameter type and `getConfig` return type? Because it shows as optional field whereas it's for sure has default value.

It's not clean to do the the following:

```ts
const width = config?.pie?.useWidth;
```

**Comment by sidharthv96** [comment]
Yes, getConfig should not have optionals. But we have to make sure to do it properly (without forced type assertion if possible), so that we get compile time errors if there is no default value for a config.

**Comment by Yokozuna59** [comment]
> Yes, getConfig should not have optionals. But we have to make sure to do it properly (without forced type assertion if possible), so that we get compile time errors if there is no default value for a config.

One possible approach is using `Required` type, something like this:

```ts
export const DEFAULT_PIE_CONFIG: Required<PieDiagramConfig> = {
   useMaxWidth: true,
   useWidth: 1200,
   textPosition: 0.75,
 } as const;
```

By doing this, it well force to set all values of the type, even optional one. Theoretically, it will always shows that it defined. And we could use `RequiredDeep` for object types from [`type-fest`](https://github.com/sindresorhus/type-fest/blob/main/source/required-deep.d.ts)

What do you think? I'll update the top message with suggested stuff in here https://github.com/mermaid-js/mermaid/issues/4499#issuecomment-1595189472 with this if all of them have been approved.

**Comment by Yokozuna59** [comment]
It would be great if we standardized diagram keywords, because some diagrams just have the actual keyword:

- pie
- timeline
- gantt
- mindmap
...etc.

And some have additional word:

- quadrantChart
- stateDiagram
- erDiagram
- sequenceDiagram
...etc.

---

I think this addition wouldn't really help; it's just making the keyword longer.

**Comment by Yokozuna59** [comment]
And maybe add `renderer` directory? 

```
packages/mermaid/src/diagrams/diagram/
├── ...
├── renderer/
│   ├── note.ts
│   ├── edge.ts
│   └── ...
└── diagramRenderer.ts # export functions from renderer dir
```

The number of files related to rendering is kinda large in number and size, specially `flowchart`, so it might be better to create a folder that contains all of those stuff. So that the diagram dir would be cleaner.

What do you think?

**Comment by sidharthv96** [comment]
> I'm aware that the root message is kind of outdated, but I haven't been receiving any feedback on the suggested ideas. Could I assume that they have been approved with those likes? @sidharthv96
> https://github.com/mermaid-js/mermaid/issues/4542#issuecomment-1609376984


Please consider the comments marked with 👍 as me agreeing personally, not as an official approval of any sorts.
We might have to make edits when others review the changes once they are inside a PR. 😄

**Comment by nirname** [comment]
Did we mention integration specs here?

**Comment by Yokozuna59** [comment]
> Did we mention integration specs here?

It might be better to mention in the script that generate new diagram, because I think this is related to definition itself, not all files related to diagram.

**Comment by nirname** [comment]
I have noticed that the name of file contains diagram name as well (`pieRenderer.ts`). It may be even better and informative, but it is also excessive a bit.

Share your thoughts on (informative, excessive)
```
packages/mermaid/src/diagrams/diagram/
├── parser/
│   └── diagram.jison
├── diagram.spec.ts
├── diagramDb.ts
├── diagramDetector.ts
├── diagramDiagram.ts
├── diagramRenderer.ts
├── diagramStyles.ts
└── diagramTypes.ts
```
versus (less informative, but very string and uniform)
```
packages/mermaid/src/diagrams/diagram/
├── parser.jison
├── tests.spec.ts
├── DB.ts
├── detector.ts
├── diagram.ts
├── renderer.ts
├── styles.ts
└── types.ts
```

**Comment by Yokozuna59** [comment]
IMO, the full form (informative) form is better because we can refer to diagram files directly rather than in the short form, and here are some reasons why:
 
In reviewing, GitHub shows and sorts files by file name, so almost all related files (to the actual diagram) will be grouped together.
Files are stored inside a 5-6 folder hierarchy, so the actual folder (diagram name) most likely won't be visible, so we won't be able to know in what diagram what file has been modified right away.
If we changed the spec folder into `tests`, we can't filter test cases easily.

But I agree with moving `jison` into diagram folder.

**Comment by Yokozuna59** [comment]
It might be better to add new folder called `diagramDefaults.ts`, this file would contain the default values for DBs, styles, themes, configs ...etc. rather than adding them to `diagramDb.ts`.

**Comment by nirname** [comment]
Didn't get your idea about defaults folder. There are too many things that can be called "default", so I'll try to sort things through:
- if we are talking about adding new diagram, then there should be a `diagramFactory` that does the trick; one of the good examples is [rails generators](https://guides.rubyonrails.org/generators.html#generator-command-line-options). **Factory would generate default structure**
- there are not so many things in diagramDB itself to be separated as a file with defaults; variables that must be initialized should stay in `clear()` function, or even better in a constructor, if DB wrapped as a class, thus `clear` probably equals to assigning new database. So **default values for DB goes to constructor**
- there is a globally defined defaultConfig, and there is https://github.com/mermaid-js/mermaid/pull/4112 that structures it which we probably should wait for, but I think that **defaultConfig must be a part of diagram**, so as to avoid extending this global config, which should be automatic composition of diagram-specific configs

**Comment by Yokozuna59** [comment]
> Didn't get your idea about defaults folder.

Just a typo :), my intention was a file. A file within diagram folder.

**Comment by nirname** [comment]
Fully agree with that

**Comment by sidharthv96** [comment]
<img width="639" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/bff32fc6-e430-464a-97e1-e070ce6ceaad">

I am aligning closer to the naming convention suggested by @nirname. Our editors are smart enough to detect `pie/db` if we type `piedb` to open a file. GitHub shows the full path in reviews. So I don't really see why we should keep the `diagram/diagramDB.ts` convention instead of `diagram/DB.ts`.

**Comment by Yokozuna59** [comment]
Then should we rename files in #4727 for parser package?

**Comment by sidharthv96** [comment]
I don't think that's necessary right now. We can discuss and decide later and do a single PR to implement the decision.
